sqlitecluster/SQLite.h

```diff
@@ -96,32 +110,137 @@ class SQLite {
+    // we commit(), which we can't do yet because we need to send outstanding transactions first.
+    //
+    // We could grab out mutex earlier, before calling processCommand(), which would avoid this situation, but it would
+    // cause all other threads to wait for the entire duration of processCommand() in this thread, which is the sort of
+    // performance problem we're trying to avoid in the first place with parallel writes. Instead, when each thread
+    // adds new commits, it makes them available in the following lists, so that we'll have access to them in
+    // sendOutstandingTransactions(), even if the current thread is operating on an old DB snapshot.
+
+    // NOTE: Both of the following collections (_inFlightTransactions and _committedtransactionIDs) are shared between
+    // all threads and need to be accessed in a synchronized fashion. They do *NOT* implement their own synchronization
+    // and must be protected by locking `_commitLock`.
+
+    // This is a map of all currently "in flight" transactions. These are transactions for which a `prepare()` has been
+    // called to generate a journal row, but have not yet been sent to peers.
+    static map<uint64_t, pair<string, string>> _inFlightTransactions;
+
+    // This is a set of transactions IDs that have been successfully committed to the database, but not yet sent to
+    // peers.
+    static set<uint64_t> _committedtransactionIDs;
+
+    // This is true when we have a transaction that we haven't committed, but does show up in the journal table. This
+    // lets us report an accurate commit count, because sometimes it's one less than the highest entry in the table.
+    // TODO: We might be able to re-use a different var for this purpose.
+    bool _uncommittedTransaction = false;
     bool _lastWriteChanged;
-    bool _readOnly;
     uint64_t _beginElapsed;
     uint64_t _readElapsed;
     uint64_t _writeElapsed;
     uint64_t _prepareElapsed;
     uint64_t _commitElapsed;
     uint64_t _rollbackElapsed;
 
+    // The name of the journal table, computed from the 'journalTable' parameter passed to our constructor.
+    string _journalName;
+
+    // A list of all the journal tables names.
+    list<string> _allJournalNames;
+
     // Sets all the pragmas we want on the provided db handle.
     void _setPragmas(sqlite3* db, int autoCheckpoint);
 
     // You're only supposed to configure SQLite options before initializing the library, but the library provides no
     // way to check if it's been initialized, so we store our own value here.
     static bool sqliteInitialized;
+
+    static string getJournalTableName(int journalTableID);
+
+    // We have designed this so that multiple threads can write to multiple journals simultaneously, but we want
+    // monotonically increasing commit numbers, so we implement a lock around changing that value.
+    static recursive_mutex _commitLock;
+    static recursive_mutex _hashLock;
```

Comment:
I think `_lastCommittedHashLock` might be more intuitive to show the relationship, and perhaps defined next to the variable it's protecting.